File: LC 2386.cpp
class Solution {
    using ll = long long;
public:
    long long kSum(vector<int>& nums, int k) {
        int n = nums.size();
        ll sum = 0;
        for (int& x : nums) {
            if (x >= 0) {
                sum += x;
            } else {
                x = -x;
            }
        }

        ranges::sort(nums);
        priority_queue<pair<ll, int>, vector<pair<ll, int>>, greater<>> pq;
        pq.emplace(nums[0], 0);
        ll ans = sum;
        for (int i = 1; i < k; ++i) {
            auto [s, j] = pq.top();
            pq.pop();
            ans = sum - s;
            if (j + 1 == n)
                continue;
            pq.emplace(s + nums[j+1], j + 1);
            pq.emplace(s + nums[j+1] - nums[j], j + 1);
        }
        return ans;
    }
};

File: LC 2386.go
func kSum(nums []int, k int) int64 {
    sum, total := 0, 0
    n := len(nums)
    for i, x := range nums {
        if x > 0 {
            sum += x
            total += x
        } else {
            total -= x
            nums[i] = -x
        }
    }
    slices.Sort(nums)

    k_largest := sort.Search(total, func(limit int) bool {
        cnt := 1
        var dfs func(int, int)
        dfs = func(i, s int) {
            if cnt == k || i == n || s + nums[i] > limit {
                return
            }
            cnt++
            dfs(i + 1, s + nums[i])
            dfs(i + 1, s)
        }
        dfs(0, 0)
        return cnt == k
    })
    return int64(sum - k_largest)
}